### PR TITLE
refactor: configmanager support [[]] format

### PR DESCRIPTION
**Pull request by steinliber**
Signed-off-by: Meng JiaFeng <jiafeng.meng@merico.dev>

## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [x] I have added relevant tests

## Description
- `configmanager` support `[[]]` format of variable
- use `yamlPath` to get yaml content
- add more `ut` for `configmanager`
 
## Related Issues
#1239

### CODE REVIEW

File: internal/pkg/configmanager/app.go [hunk 1]
```diff
@@ -1,84 +1,118 @@
...
-	// which will be rendered to "CIPipelines" field in App with "PipelineTemplates".
-	// The "CDRawConfigs" is similar to "CIRawConfigs".
-	AppInConfig struct {
-		Name         string        `yaml:"name" mapstructure:"name"`
-		Spec         RawOptions    `yaml:"spec" mapstructure:"spec"`
-		Repo         *Repo         `yaml:"repo" mapstructure:"repo"`
-		RepoTemplate *RepoTemplate `yaml:"repoTemplate" mapstructure:"repoTemplate"`
-		CIRawConfigs []CICD        `yaml:"ci" mapstructure:"ci"`
-		CDRawConfigs []CICD        `yaml:"cd" mapstructure:"cd"`
+// getAppTools return app tools
```
Comment (IronCore864) [inline internal/pkg/configmanager/app.go:28]: getAppTools is not very descriptive. I suggest a better name. For example, getToolsFromApp?

File: internal/pkg/configmanager/app.go [hunk 2]
```diff
@@ -1,84 +1,118 @@
...
-)
+	rawData.setDefault()
+	appVars := mapz.MergeMaps(globalVars, rawData.Spec)
+	// 3. generate app repo and tempalte repo from scmInfo
+	repoScaffoldingTool, err := rawData.getRepoTemplateTool(appVars)
+	if err != nil {
+		return nil, fmt.Errorf("app[%s] get repo failed: %w", rawData.Name, err)
+	}
+	// 4. get ci/cd pipelineTemplates
+	tools, err := rawData.getPipelineTools(templateMap, appVars)
```
Comment (IronCore864) [inline internal/pkg/configmanager/app.go:49]: Same as above; generateCICDToolsFromAppConfig? In this way it's consistent with the comment at line 59.

File: internal/pkg/configmanager/app.go [hunk 3]
```diff
@@ -1,84 +1,118 @@
...
 	}
-	return errs
+	return tools, nil
 }
 
-func (app *App) validate() error {
-	if app.Repo.RepoInfo != nil && app.Repo.RepoInfo.Name == "" {
-		app.Repo.RepoInfo.Name = app.Name
+// getRepoTemplateTool will use repo-scaffolding plugin for app
+func (a *appRaw) getRepoTemplateTool(appVars map[string]any) (*Tool, error) {
```
Comment (IronCore864) [inline internal/pkg/configmanager/app.go:79]: getRepoScaffoldingTool?

File: internal/pkg/configmanager/app.go [hunk 4]
```diff
@@ -1,84 +1,118 @@
...
+// setDefault will set repoName to appName if repo.name field is empty
+func (a *appRaw) setDefault() {
+	if a.Repo != nil && a.Repo.Name == "" {
+		a.Repo.Name = a.Name
 	}
+}
 
-	return configFinal
+// since all plugin depends on code is deployed, get dependsOn for repoTemplate
+func (a *appRaw) getRepoTemplateDepends() []string {
```
Comment (IronCore864) [inline internal/pkg/configmanager/app.go:111]: getRepoScaffoldingDependants

File: internal/pkg/configmanager/config.go [hunk 5]
```diff
@@ -1,67 +1,255 @@
...
+	// Config records rendered config values and is used as a general config in DevStream.
+	Config struct {
+		// Command line flag have a higher priority than the config file.
+		// If you used the `--plugin-dir` flag with `dtm`, then the "pluginDir" in the config file will be ignored.
+		PluginDir string
+		Tools     Tools `yaml:"tools"`
+		State     *State
+	}
+	// ConfigRaw is used to describe original raw configs read from files
+	ConfigRaw struct {
```
Comment (IronCore864) [inline internal/pkg/configmanager/config.go:31]: How about RawConfig? Sounds more intuitive.

File: pkg/util/file/file.go [hunk 6]
```diff
@@ -42,3 +44,37 @@ func CopyFile(srcFile, dstFile string) (err error) {
 	}
 	return dFile.Sync()
 }
+
+// GenAbsFilePath return all the path with a given file name
+func GenAbsFilePath(baseDir, file string) (string, error) {
```
Comment (IronCore864) [inline pkg/util/file/file.go:49]: Gen -> generate

Gen could also mean generic?

Function names can be longer if they improve readability.

File: pkg/util/mapz/map.go [hunk 7]
```diff
@@ -23,3 +24,11 @@ func DecodeStructToMap(structVars any) (map[string]interface{}, error) {
 	}
 	return rawConfigVars, nil
 }
+
+// MergeMaps merge two maps
+// if there are same keys in two maps, the key of second one will overwrite the first one
+func MergeMaps(m1 map[string]any, m2 map[string]any) map[string]any {
```
Comment (IronCore864) [inline pkg/util/mapz/map.go:30]: When calling this function, mapz.MergeMaps(m1, m2) contains redundant information(mapz, merge maps)

mapz.Merge(m1, m2) could be simpler without losing the meaning.

**Review by IronCore864 — CHANGES_REQUESTED** [review]

**Review by daniel-hutao — APPROVED** [review]

**Review by IronCore864 — APPROVED** [review]

**Pull request merged**